rust-alternate/src/connection.rs

```diff
@@ -0,0 +1,256 @@
+    type Error = anyhow::Error;
+
+    fn poll_ready(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        let this = self.project();
+        let stream = this.write;
+
+        stream.poll_ready(cx)
+    }
+
+    fn start_send(self: Pin<&mut Self>, message: Message) -> Result<(), Self::Error> {
+        let this = self.project();
+        let stream = this.write;
+
+        stream.start_send(message)
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        let this = self.project();
+        let stream = this.write;
+
+        stream.poll_flush(cx)
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        let this = self.project();
+        let stream = this.write;
+
+        stream.poll_close(cx)
+    }
+}
+
+impl ConnectionReadHalf {
+    /// Returns the socket address of the connection
+    pub fn get_addr(&self) -> SocketAddr {
+        self.addr
+    }
+
+    /// Waits for `ReadyUp` message from the connection
+    pub async fn wait_ready_up(&mut self) -> Result<()> {
+        let message = self.receive_message().await?;
+
+        if matches!(message.message_type, MessageType::ReadyUp) {
+            Ok(())
+        } else {
+            Err(anyhow::anyhow!("expected `Play` message"))
+        }
+    }
+
+    /// Receives a message from websocket connection
+    async fn receive_message(&mut self) -> Result<Message> {
+        self.next().await.context(format!(
+            "unable to receive websocket message from player {}",
+            self.addr
+        ))?
+    }
+}
+
+impl ConnectionWriteHalf {
+    /// Returns the socket address of the connection
+    pub fn get_addr(&self) -> SocketAddr {
+        self.addr
+    }
+
+    /// Sends `ReadyUp` message to the connection
+    pub async fn ready_up(&mut self) -> Result<()> {
+        self.send_message(MessageType::ReadyUp).await
+    }
+
+    /// Sends `Play` message to the connection
+    pub async fn play(&mut self) -> Result<()> {
+        self.send_message(MessageType::Play).await
+    }
+
+    /// Sends a `GameOver` message to the connection with data as `loser`
+    pub async fn send_loser(&mut self) -> Result<()> {
+        self.send_message(Message::new(MessageType::GameOver).with_data("loser".to_owned()))
+            .await
+    }
+
+    /// Sends a `GameOver` message to the connection with data as `winner`
+    pub async fn send_winner(&mut self, game_stats: &GameStats, active_games: usize) -> Result<()> {
+        let message_str = format!("winner({})___{}", active_games, game_stats);
+
+        self.send_message(Message::new(MessageType::GameOver).with_data(message_str))
+            .await
+    }
+
+    /// Closes the connection
+    pub async fn close(mut self) -> Result<()> {
+        self.sink
+            .close()
+            .await
+            .context("failed to close connection")
```

Comment:
I don't know much about this `context`.  Looks like expect, but it must be used for something, wanna give me the high level view of why its nice to use it (in your perspective)?

Reply:
.context is like explicit building of traces

You may treat this as
```rs
.map_err(|err| format!("{err}\n{context}"))
```

Or
```go
errors.Wrap(err, context);
```

Reply:
I usually use a custom error type with thisError, but if you aren't doing that this isn't a bad way to provide context for your errors.

Reply:
i have never played with anyhow, but it seems pretty neet.

Reply:
`anyhow` and `thiserror`, both are developed by the same author. `thiserror` is more useful for library crates and `anyhow` is easier to use for applications. `anyhow` provides a quick way of adding context to the error and you don't have to create your own error type.